### ISSUE TITLE
An option to add automatic IDs to headings

### ISSUE THREAD

**Issue by Wilfred** [atlassian/commonmark-java#26]
GitHub flavoured markdown adds automatic IDs to headings, and it would be great to see this as an option here (disclaimer: I'm hoping to see the feature in Stash/Bitbucket server).

This is not in commonmark ([and won't be added](http://talk.commonmark.org/t/feature-request-automatically-generated-ids-for-headers/115)) though [some implementations even do this by default](http://johnmacfarlane.net/babelmark2/?text=%23+Heading).

In the case of GitHub, a heading `#foo` gets an ID `user-generated-foo` to avoid clobbering IDs used elsewhere. There's then some JS that means that you can use `#foo` in your URL instead of `#user-generated-foo`.

This would allow us to use tools like [markdown-toc](https://github.com/jonschlinkert/markdown-toc) or [doctoc](https://github.com/thlorenz/doctoc) on Bitbucket server.

**Comment by robinst** [comment]
Cheeky :). But I agree, this would be useful.

Should be pretty simple to build it as an extension using an [`AttributeProvider`](https://github.com/atlassian/commonmark-java/blob/master/commonmark/src/main/java/org/commonmark/html/AttributeProvider.java) that looks at `Header` nodes.

I won't have time in the next few weeks, but if someone else wants to take a stab at this, you're welcome.

**Comment by robinst** [comment]
I've looked into what existing implementations do for converting the heading text to an ID:

http://johnmacfarlane.net/babelmark2/?text=%23+Heading%0A%0A%23+heading%0A%0A%23+heading%252%0A%0A%23+heading{}3%0A%0A%23+heading+%C3%A4%0A%0A%23+heading+%F0%9F%98%84%0A%0A%23+heading+%F0%9F%98%84%0A%0A%23+heading+%F0%9F%98%84%0A%0A%23+heading++4%0A%0A%23+1+heading%0A%0A%23+123%0A%0A%23+--%0A%0A%23+00A1%C2%A1test%0A%0A%23+A2%C2%A2test

HTML itself doesn't put any restrictions on the format, but other specifications do. One that is probably useful to make IDs conform to is CSS:

http://www.w3.org/TR/CSS21/syndata.html#value-def-identifier

Looks like that is what Pandoc does, except that it strips more characters such as ¢ (U+00A2).

It also avoids duplicates by appending `-N`.

**Comment by vsch** [comment]
Here is the implementation I use to replicate GitHub behaviour. It is Kotlin but should be quite readable to Java users.

The header text is mapped as follows:
- spaces and dash `-` map to dash `-`
- underscore unchanged
- alpha to lowercase, It is **important** to handle non-ASCII alpha characters properly. I didn't at first and this resulted in headers that contained all Chinese characters to have no ids. 
- numeric unchanged
- every thing else skipped

If an id was already encountered in a top-down page parse then `-#` is appended to the id, where # is an incremented integer, starting with 1.

No attempt is made to disambiguate id's appearing further down the page. So `# abc` `# abc` and `# abc-1` will produce three headers with the following ids: `abc`, `abc-1`, `abc-1`. This is how GitHub does it, last I checked.
 
It sounds like the `AttributeProvider` is the right approach. I implemented something similar in my pegdown fork to do this. My use of the parser would require many user selectable options that change the parser and HTML generator flavours to match the algorithm of the target markdown processor. 

In [idea-multimarkdown plugin](https://github.com/vsch/idea-multimarkdown) I implemented inspections that validate links and anchor refs for being undefined or having duplicates. I also have anchor ref completions and refactoring so it is important for me to have the same behaviour as GitHub, otherwise inspections would be meaningless and completions/refactoring would result in erroneous anchor refs.

```kotlin
class GitHubHeaderIdComputer: HeaderIdComputer {
    protected val headerBaseIds = HashMap<String, Int>()

    override fun computeHeaderId(node: HeaderNode, anchorLinkNode: AnchorLinkNode?, headerText: String): String {
        var baseRefId = computeHeaderId(headerText)

        if (headerBaseIds.containsKey(baseRefId)) {
            var index = headerBaseIds[baseRefId] as Int
            index++
            headerBaseIds.put(baseRefId, index);
            baseRefId += "-$index"
        } else {
            headerBaseIds.put(baseRefId, 0)
        }

        return baseRefId
    }

    companion object {
        fun computeHeaderId(headerText: String): String {
            var baseRefId = ""
            for (c in headerText) {
                when (c) {
                    ' ' -> baseRefId += '-'
                    '-' -> baseRefId += '-'
                    '_' -> baseRefId += '_'
                    else -> {
                        if ( c.isAlphaNumeric()) baseRefId += toLowerCase(c)
                    }
                }
            }
            return baseRefId
        }
    }
}
```

**Comment by robinst** [comment]
Thanks for the insights, @vsch.

Just released 0.7.0 with an extension for this, thanks @paulthom12345 🎉: https://github.com/atlassian/commonmark-java/releases/tag/commonmark-parent-0.7.0
